src/core/qgspointlocator.cpp

```diff
@@ -332,195 +332,39 @@ static QgsPointLocator::MatchList _geometrySegmentsInRect( QgsGeometry *geom, co
-    return lst;
-
-  _CohenSutherland cs( rect );
-
-  QgsConstWkbPtr wkbPtr( wkb );
-  wkbPtr.readHeader();
-
-  QgsWkbTypes::Type wkbType = geom->wkbType();
-
-  bool hasZValue = false;
-  switch ( wkbType )
-  {
-    case QgsWkbTypes::Point25D:
-    case QgsWkbTypes::Point:
-    case QgsWkbTypes::MultiPoint25D:
-    case QgsWkbTypes::MultiPoint:
-    {
-      // Points have no lines
-      return lst;
-    }
-
-    case QgsWkbTypes::LineString25D:
-      hasZValue = true;
-      //intentional fall-through
-      FALLTHROUGH
-    case QgsWkbTypes::LineString:
-    {
-      int nPoints;
-      wkbPtr >> nPoints;
-
-      double prevx = 0.0, prevy = 0.0;
-      for ( int index = 0; index < nPoints; ++index )
-      {
-        double thisx = 0.0, thisy = 0.0;
-        wkbPtr >> thisx >> thisy;
-        if ( hasZValue )
-          wkbPtr += sizeof( double );
-
-        if ( index > 0 )
-        {
-          if ( cs.isSegmentInRect( prevx, prevy, thisx, thisy ) )
-          {
-            QgsPointXY edgePoints[2];
-            edgePoints[0].set( prevx, prevy );
-            edgePoints[1].set( thisx, thisy );
-            lst << QgsPointLocator::Match( QgsPointLocator::Edge, vl, fid, 0, QgsPointXY(), index - 1, edgePoints );
-          }
-        }
-
-        prevx = thisx;
-        prevy = thisy;
-      }
-      break;
-    }
-
-    case QgsWkbTypes::MultiLineString25D:
-      hasZValue = true;
-      //intentional fall-through
-      FALLTHROUGH
-    case QgsWkbTypes::MultiLineString:
-    {
-      int nLines;
-      wkbPtr >> nLines;
-      for ( int linenr = 0, pointIndex = 0; linenr < nLines; ++linenr )
-      {
-        wkbPtr.readHeader();
-        int nPoints;
-        wkbPtr >> nPoints;
 
-        double prevx = 0.0, prevy = 0.0;
-        for ( int pointnr = 0; pointnr < nPoints; ++pointnr )
-        {
-          double thisx = 0.0, thisy = 0.0;
-          wkbPtr >> thisx >> thisy;
-          if ( hasZValue )
-            wkbPtr += sizeof( double );
-
-          if ( pointnr > 0 )
-          {
-            if ( cs.isSegmentInRect( prevx, prevy, thisx, thisy ) )
-            {
-              QgsPointXY edgePoints[2];
-              edgePoints[0].set( prevx, prevy );
-              edgePoints[1].set( thisx, thisy );
-              lst << QgsPointLocator::Match( QgsPointLocator::Edge, vl, fid, 0, QgsPointXY(), pointIndex - 1, edgePoints );
-            }
-          }
-
-          prevx = thisx;
-          prevy = thisy;
-          ++pointIndex;
-        }
-      }
-      break;
-    }
+  QgsGeometry straightGeom = geom->convertToType( QgsWkbTypes::LineGeometry, true );
+  straightGeom.convertToStraightSegment();
 
-    case QgsWkbTypes::Polygon25D:
-      hasZValue = true;
-      //intentional fall-through
-      FALLTHROUGH
-    case QgsWkbTypes::Polygon:
-    {
-      int nRings;
-      wkbPtr >> nRings;
+  // A linestring should/(must?) have at least two points
+  // TODO: This is a workaround to be fixed in Curve types
```

Comment:
My assumption is that this is the correct behaviour and should simply be commented as "checking for invalid linestrings" and that we won't be able to secure the whole system against invalid geometries without breaking a lot of code and workflows.